Communications hg

hgrehm opened the pull request.

hgrehm committed:
Changed GCS to server and drone to client

hgrehm committed:
Added default addr for RCTTCPClient

hgrehm committed:
Set server listening addr to ''

melinadika committed:
Added support for cones

melinadika committed:
Rearranged functions to resolve errors

melinadika committed:
Callable to handle disconnect event

hgrehm committed:
Separated Server and ServerConnection; began converting gcsComms

hgrehm committed:
Merge branch 'communications_HG' into communications_HG

melinadika committed:
Allowing multiple clients to connect

hgrehm committed:
Caught and resolved minor errors which arose when running the simulator

hgrehm committed:
Caught and resolved minor errors which arose when running the simulator

melinadika committed:
move stcomms into repository

melinadika committed:
removed unnecessary functionality from unused implementation for multiple clients

melinadika committed:
integrating sleep timer comms and changes for multiple clients in RCT comms

ntlhui committed:
Fixed alarm definitions

ntlhui committed:
Added sleep timer test

ntlhui committed:
Updated sleep timer

hgrehm committed:
Merge branch 'communications_HG' of https://github.com/UCSD-E4E/radio_collar_tracker_comms into communications_HG

ntlhui committed:
Added disconnect behavior setter

ntlhui committed:
Updated version

ntlhui committed:
Added package version

ntlhui committed:
Update version

hgrehm committed:
Merge branch 'master' into communications_HG

ntlhui commented on RCTComms/comms.py:
```diff
@@ -871,8 +910,9 @@ class gcsComms:
     Radio Collar Tracker UDP Interface
     '''
     __BUFFER_LEN = 1024
+    __lock = threading.Lock()
```
Should this be a static class scope variable?  Or should it be an instance variable, defined in `__init__`?

ntlhui commented on RCTComms/comms.py:
```diff
@@ -1022,11 +1078,18 @@ def registerCallback(self, event: EVENTS, callback: Callable) -> None:
         else:
             self.__packetMap[event.value] = [callback]
 
+    def synchronizedCallback(self, callback):
+        def lockAndCall(packet: rctBinaryPacket, addr: str):
+            with gcsComms.__lock:
+                callback(packet, addr)
+        return lockAndCall
+
```
Should this be a wrapper, or should this logic be embedded into the callback logic? i.e. in comms.py:1006
Also, what specifically are we locking access to?

hgrehm replied:
Re: locking access
I believe the lock is there to prevent us from somehow sending packets out of order, since we create a new GCSComms object per connection in tower mode and each one spins its own receiver thread. That's why it's a class variable as well (one lock for all the GCSCmms objects). I tried removing all lock functionality to confirm, and the simulator crashed within a few seconds of starting when I ran multiple clients. I can embed the lock into the callback logic though.

ntlhui commented on RCTComms/comms.py:
```diff
@@ -1159,4 +1225,7 @@ def execute_cb(self, event_value: int, kwargs):
             cb_(**kwargs)
 
     def registerCallback(self, event: EVENTS, callback):
-        self.__packetMap[event.value].append(callback)
+        if event.value in self.__packetMap:
+            self.__packetMap[event.value].append(callback)
+        else:
+            self.__packetMap[event.value] = [callback]
```
It may be better instead to instantiate `self.__packetMap` with empty lists to start with

ntlhui reviewed (changes requested):
Please update all tests to reflect the new usage

hgrehm committed:
Merge branch 'master' into communications_HG

hgrehm committed:
Updated tests to reflect usage post-merge

Wesleylewis05 committed:
small temp fix for line 265 bug (changed break to pass)

Commit:
Added coverage and wheel

hgrehm committed:
Increased time limit for testing data transport to avoid timeout

hgrehm committed:
Updated option names to snake case to support gcs style updates

ntlhui commented on RCTComms/comms.py:
```diff
@@ -871,8 +910,9 @@ class gcsComms:
+    __lock = threading.Lock()
 
-    def __init__(self, port: RCTComms.transport.RCTAbstractTransport, GC_HeartbeatWatchdogTime=30):
+    def __init__(self, port: RCTComms.transport.RCTAbstractTransport, disconnected: Callable[[], None], GC_HeartbeatWatchdogTime=30):
```
The `disconnected` parameter should probably have a default value of `None` and type `Optional[Callable[[], None]`

ntlhui commented on RCTComms/comms.py:
```diff
@@ -1022,10 +1076,14 @@ def registerCallback(self, event: EVENTS, callback: Callable) -> None:
+        def lockAndCall(packet: rctBinaryPacket, addr: str):
+            with gcsComms.__lock:
+                callback(packet, addr)
+        return lockAndCall
```
This should probably be a private or hidden method - there is no reason for someone outside of this object to call synchronization

ntlhui commented on RCTComms/comms.py:
```diff
@@ -1043,7 +1101,7 @@ def sendMessage(self, payload: bytes, packetClass: int, packetID: int):
         cksum = binascii.crc_hqx(msg, 0xFFFF).to_bytes(2, 'big')
         self.__log.info("Send: %s" % ((msg + cksum).hex()))
-        self.sock.send(msg, self.__mavIP)
+        self.sock.send(msg)
```
If we intend to drop the second argument, we should probably make a structural change to `RCTAbstractTransport` so that `RCTAbstractTransport.send` only takes a single argument.

ntlhui commented on RCTComms/transport.py:
```diff
@@ -34,13 +39,17 @@
+import threading
+from typing import Callable, Optional, Tuple
 import logging
 import time
```
Please sort includes

ntlhui commented on RCTComms/transport.py:
```diff
@@ -222,7 +231,7 @@ def isOpen(self):
 
 class RCTTCPClient(RCTAbstractTransport):
-    def __init__(self, port: int, addr: str):
+    def __init__(self, port: int=9000, addr: str='255.255.255.255'):
```
I'm not sure I agree with the TCP client having default values - the intent behind the client is to connect to a server.  There should not be a notion of a default server to connect to.

ntlhui commented on RCTComms/transport.py:
```diff
@@ -238,74 +247,163 @@ def close(self):
+            if e.winerror == 10038:
+                pass
+            else:
+                raise
```
You may want to add comments to explain why we are specifically "ignoring" the 10038 error.

ntlhui commented on RCTComms/transport.py:
```diff
@@ -238,74 +247,163 @@ def close(self):
+        for new connections.
+        '''
+        # Use printed addr in rctconfig
+        print ('Server started at {}'.format(socket.gethostbyname(socket.gethostname())))
```
Let's use logging facilities for this instead

ntlhui commented on RCTComms/transport.py:
```diff
@@ -238,74 +247,163 @@ def close(self):
-        self.__conn: Optional[socket.socket] = None
-        self.__addr: Optional[Tuple[str, int]] = None
+        self.__generatorThread: Optional[threading.Thread] = None
+        self.running = False
```
Consider using `threading.Event` for this instead.

ntlhui commented on RCTComms/transport.py:
```diff
@@ -238,74 +247,163 @@ def close(self):
+    def __init__(self, addr: Tuple[str, int], conn: socket.socket, id: int):
+        self.__addr = addr
+        self.__socket = conn
+        self.running = False
```
This doesn't seem to do anything...

ntlhui commented on RCTComms/transport.py:
```diff
@@ -238,74 +247,163 @@ def close(self):
+
+        while self.running:
+            try:
+                clientConn, clientAddr = self.__socket.accept()
```
This action will block this entire loop.  `__generatorThread.join` will not cause this to exit.  As a result, `__socket.close` will never execute, and you will have hanging resources.

One option is to make generatorThread a daemon.  Other option is to add `self.running` as an Event to a selector

ntlhui reviewed (changes requested):
Please address comments.

Note that current tests do not cover UDP send/receive, rctConePacket

hgrehm committed:
added timeout to TCP Server's socket and UDP to data transport test

hgrehm commented:
Requested items have been addressed:
- UDP added to transport test, currently not included in comms test (can be addressed in the future)
- disconnected parameter type and default value updated
- privatized synchronizedCallback
- restored self.__mavIP to send methods and ignored unused argument
- sorted includes
- removed catch of windows error--this was related to the unclosed socket from RCTTCPServer (see final bullet)
- replaced print statement with log statement
- changed self.running to an event
- removed RCTTCPServerConnection's pointless running variable
- gave RCTTCPServer a timeout of 2s, so that the client-accepting loop can actually stop once running event is set 
and socket can close; switching __generatorThread to a daemon didn't allow us to properly close the socket, even though the thread did get stopped, hence the timeout

Commit:
Reenabling tests

Commit:
Increase test_open timeout to 20
